File: app/src/main/java/org/autojs/autojs/ui/main/drawer/DrawerMenuItemViewHolder.java
package org.autojs.autojs.ui.main.drawer;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import org.autojs.autojs.ui.widget.BindableViewHolder;
import org.autojs.autojs.ui.widget.PrefSwitch;
import org.autojs.autojs.ui.widget.SwitchCompat;
import org.autojs.autojs6.R;

import java.io.IOException;

import butterknife.BindView;
import butterknife.ButterKnife;
import me.zhanghai.android.materialprogressbar.MaterialProgressBar;

/**
 * Created by Stardust on 2017/12/10.
 */
public class DrawerMenuItemViewHolder extends BindableViewHolder<DrawerMenuItem> {

    public static final long CLICK_TIMEOUT = 540;
    @BindView(R.id.sw)
    PrefSwitch mSwitchCompat;

    @BindView(R.id.progress_bar)
    MaterialProgressBar mProgressBar;

    @BindView(R.id.icon)
    ImageView mIcon;

    @BindView(R.id.title)
    TextView mTitle;

    @BindView(R.id.subtitle)
    TextView mSubtitle;

    private boolean mAntiShake;
    private long mLastClickMillis;
    private DrawerMenuItem mDrawerMenuItem;

    public DrawerMenuItemViewHolder(View itemView) {
        super(itemView);
        ButterKnife.bind(this, itemView);
        mSwitchCompat.setOnCheckedChangeListener((buttonView, isChecked) -> {
            try {
                onClick();
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
        itemView.setOnClickListener(v -> {
            if (mSwitchCompat.getVisibility() == VISIBLE) {
                mSwitchCompat.toggle();
            } else {
                try {
                    onClick();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        });
    }

    @Override
    public void bind(DrawerMenuItem item, int position) {
        mDrawerMenuItem = item;
        mIcon.setImageResource(item.getIcon());
        mTitle.setText(item.getTitle());

        String subtitle = item.getSubtitle();
        mSubtitle.setText(subtitle);
        mSubtitle.setVisibility(subtitle != null ? VISIBLE : GONE);

        mAntiShake = item.antiShake();
        setSwitch(item);
        setProgress(item.isProgress());
    }

    private void setSwitch(DrawerMenuItem item) {
        if (!item.isSwitchEnabled()) {
            mSwitchCompat.setVisibility(GONE);
            return;
        }
        mSwitchCompat.setVisibility(VISIBLE);
        int prefKey = item.getPrefKey();
        if (prefKey == DrawerMenuItem.DEFAULT_PREFERENCE_KEY) {
            mSwitchCompat.setChecked(item.isChecked(), false);
            mSwitchCompat.setPrefKey(null);
        } else {
            mSwitchCompat.setPrefKey(itemView.getResources().getString(prefKey));
        }
    }

    private void onClick() throws IOException {
        mDrawerMenuItem.setChecked(mSwitchCompat.isChecked());
        if (mAntiShake && (System.currentTimeMillis() - mLastClickMillis < CLICK_TIMEOUT)) {
            mSwitchCompat.setChecked(!mSwitchCompat.isChecked(), false);
            return;
        }
        mLastClickMillis = System.currentTimeMillis();
        if (mDrawerMenuItem != null) {
            mDrawerMenuItem.performAction(this);
        }
    }

    private void setProgress(boolean onProgress) {
        mProgressBar.setVisibility(onProgress ? VISIBLE : GONE);
        mIcon.setVisibility(onProgress ? GONE : VISIBLE);
        mSwitchCompat.setEnabled(!onProgress);
        itemView.setEnabled(!onProgress);
    }

    public SwitchCompat getSwitchCompat() {
        return mSwitchCompat;
    }

}

File: app/src/main/java/org/autojs/autojs/ui/main/drawer/DrawerMenuItemViewHolder.kt
package org.autojs.autojs.ui.main.drawer

import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.core.view.isVisible
import me.zhanghai.android.materialprogressbar.MaterialProgressBar
import org.autojs.autojs.ui.widget.BindableViewHolder
import org.autojs.autojs.ui.widget.PrefSwitch
import org.autojs.autojs.util.ViewUtils.installFullRowRippleForwarder
import org.autojs.autojs6.databinding.DrawerMenuItemBinding

/**
 * Created by Stardust on Dec 10, 2017.
 * Modified by SuperMonster003 as of Jan 14, 2026.
 * Transformed by SuperMonster003 on May 13, 2023.
 */
class DrawerMenuItemViewHolder(itemView: View) : BindableViewHolder<DrawerMenuItem?>(itemView) {

    private val mSwitchCompat: PrefSwitch
    private val mSwitchDivider: View
    private val mProgressBar: MaterialProgressBar
    private val mIcon: ImageView
    private val mTitle: TextView
    private val mSubtitle: TextView

    private var mAntiShake = false
    private var mLastClickMillis: Long = 0

    private lateinit var mDrawerMenuItem: DrawerMenuItem

    init {
        val binding = DrawerMenuItemBinding.bind(itemView)
        mSwitchCompat = binding.sw
        mSwitchDivider = binding.switchDivider
        mProgressBar = binding.progressBar
        mIcon = binding.icon
        mTitle = binding.title
        mSubtitle = binding.subtitle
        mSwitchCompat.setOnCheckedChangeListener { _, _ ->
            onClick()
        }

        val iconContainer = binding.iconContainer
        val titleContainer = binding.titleContainer

        // Install full-row ripple forwarding once.
        // zh-CN: 安装整行 Ripple 转发, 通常只需安装一次.
        installFullRowRippleForwarder(itemView, iconContainer, titleContainer)

        val toggleAction = {
            if (mSwitchCompat.isVisible) {
                mSwitchCompat.toggle()
            } else {
                onClick()
            }
        }

        iconContainer.setOnClickListener {
            toggleAction()
        }

        iconContainer.setOnLongClickListener {
            when (val menuItem = mDrawerMenuItem) {
                is DrawerMenuToggleableItem -> {
                    menuItem.launchManagerIfPossible()
                }
                else -> false
            }
        }

        titleContainer.setOnClickListener {
            when (val menuItem = mDrawerMenuItem) {
                is DrawerMenuToggleableItem -> {
                    menuItem.onTitleContainerClick()
                }
                else -> toggleAction()
            }
        }

        titleContainer.setOnLongClickListener {
            when (val menuItem = mDrawerMenuItem) {
                is DrawerMenuToggleableItem -> {
                    menuItem.launchManagerIfPossible()
                }
                else -> false
            }
        }
    }

    override fun bind(data: DrawerMenuItem?, position: Int) {
        data ?: return
        mDrawerMenuItem = data
        mIcon.setImageResource(data.icon)
        mTitle.setText(data.title)

        val subtitle = data.subtitle
        mSubtitle.text = subtitle
        mSubtitle.visibility = if (subtitle != null) View.VISIBLE else View.GONE

        mAntiShake = data.antiShake()
        setSwitch(data)
        setProgress(data.isProgress)
    }

    fun isChecked() = mSwitchCompat.isChecked

    private fun setSwitch(item: DrawerMenuItem) {
        if (!item.isSwitchEnabled) {
            mSwitchCompat.visibility = View.GONE
            mSwitchDivider.visibility = View.GONE
            return
        }
        mSwitchCompat.visibility = View.VISIBLE
        mSwitchDivider.visibility = View.VISIBLE
        val prefKey = item.prefKey
        if (prefKey == DrawerMenuItem.DEFAULT_PREFERENCE_KEY) {
            mSwitchCompat.setChecked(item.isChecked, false)
            mSwitchCompat.setPrefKey(null)
        } else {
            mSwitchCompat.setPrefKey(itemView.resources.getString(prefKey))
        }
    }

    private fun onClick() {
        runCatching {
            mDrawerMenuItem.isChecked = mSwitchCompat.isChecked
            if (mAntiShake && System.currentTimeMillis() - mLastClickMillis < CLICK_TIMEOUT) {
                mSwitchCompat.setChecked(!mSwitchCompat.isChecked, false)
                return
            }
            mLastClickMillis = System.currentTimeMillis()
            mDrawerMenuItem.performAction(this)
        }
    }

    private fun setProgress(onProgress: Boolean) {
        mProgressBar.visibility = if (onProgress) View.VISIBLE else View.GONE
        mIcon.visibility = if (onProgress) View.GONE else View.VISIBLE
        mSwitchCompat.isEnabled = !onProgress
        itemView.isEnabled = !onProgress
    }

    companion object {

        const val CLICK_TIMEOUT: Long = 540

    }

}
